postgres_ffi/src/xlog_utils.rs

```diff
@@ -292,3 +292,48 @@ impl XLogRecord {
         self.xl_info == pg_constants::XLOG_SWITCH && self.xl_rmid == pg_constants::RM_XLOG_ID
     }
 }
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct XLogPageHeaderData {
+    pub xlp_magic: u16,    /* magic value for correctness checks */
+    pub xlp_info: u16,     /* flag bits, see below */
+    pub xlp_tli: u32,      /* TimeLineID of first record on page */
+    pub xlp_pageaddr: u64, /* XLOG address of this page */
+    pub xlp_rem_len: u32,  /* total len of remaining data for record */
+    padding: u32,          /* Add explicit padding */
+}
+
+impl XLogPageHeaderData {
+    pub fn from_bytes<B: Buf>(buf: &mut B) -> XLogPageHeaderData {
+        XLogPageHeaderData {
+            xlp_magic: buf.get_u16_le(),
+            xlp_info: buf.get_u16_le(),
+            xlp_tli: buf.get_u32_le(),
+            xlp_pageaddr: buf.get_u64_le(),
+            xlp_rem_len: buf.get_u32_le(),
+            padding: buf.get_u32_le(),
+        }
+    }
+}
+
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct XLogLongPageHeaderData {
+    pub std: XLogPageHeaderData, /* standard header fields */
+    pub xlp_sysid: u64,          /* system identifier from pg_control */
+    pub xlp_seg_size: u32,       /* just as a cross-check */
+    pub xlp_xlog_blcksz: u32,    /* just as a cross-check */
+}
+
+impl XLogLongPageHeaderData {
+    pub fn from_bytes<B: Buf>(buf: &mut B) -> XLogLongPageHeaderData {
```

Comment:
This syntax is new to me. Is it the same as this?
```
pub fn from_bytes(buf: &mut Buf) -> ...
```

Reply:
No, it is slightly different.
`from_bytes<B: Buf>(buf: &mut B) `
this syntax allows accepting any type that implements Buf trait.

see
https://doc.rust-lang.org/rust-by-example/generics/bounds.html

It will allow using this code for both Bytes and plain slices